### PR TITLE
Fix DALI paddle plugin stream synchronization error

### PR DESCRIPTION
**Pull request by sneaxiy**
## Description
- [x] **Bug fix** (*non-breaking change which fixes an issue*)
- [ ] **New feature** (*non-breaking change which adds functionality*)
- [ ] **Breaking change** (*fix or feature that would cause existing functionality to not work as expected*)
- [ ] **Refactoring** (*Redesign of existing code that doesn't affect functionality*)
- [ ] **Other** (*e.g. Documentation, Tests, Configuration*)

### What happened in this PR

As far as I know:

- For DALI: if the given stream is `None` when calling the `dali_tensor.copy_to_external()` method, DALI would use an internal stream (non-default stream) to copy the data. Let's call this DALI internal stream to be `dali_stream`.
- For PaddlePaddle: PaddlePaddle uses its own internal stream to launch the CUDA kernel. Let's call this PaddlePaddle internal stream to be `paddle_stream`.

The code of `dali_tensor.copy_to_external(stream=None, non_blocking=False)` is something like this:
```
if stream is None:
    stream = dali_stream

cudaMemcpyAsync(external_tensor_ptr, dali_tensor_ptr, dali_stream)
if not non_blocking:
    cudaStreamSynchronize(dali_stream)
```

When calling `cudaMemcpyAsync`, the `external_tensor_ptr` may be used in the launched CUDA kernel inside the PaddlePaddle framework. Therefore, the `external_tensor_ptr` may be read using `paddle_stream` and be written using `dali_stream` in the same time. This would cause incorrect results. We found this bug when running ResNet50 model of PaddlePaddle.

This PR makes the `cudaMemcpyAsync` use the `paddle_stream` instead of `dali_stream` to make sure the proper synchronization.

#### Additional information
- Affected modules and functionalities: DALI PaddlePaddle plugin

- Key points relevant for the review: CUDA stream synchronization between the DALI stream and PaddlePaddle stream

### Tests
- [x] Existing tests apply
- [ ] New tests added
  - [ ] Python tests
  - [ ] GTests
  - [ ] Benchmark
  - [ ] Other
- [ ] N/A

### Documentation
- [x] Existing documentation applies
- [ ] Documentation updated
  - [ ] Docstring
  - [ ] Doxygen
  - [ ] RST
  - [ ] Jupyter
  - [ ] Other
- [ ] N/A

#### Requirements
- [ ] Implements new requirements
- [ ] Affects existing requirements
- [ ] N/A

**REQ IDs**: N/A

**JIRA TASK**: N/A

### CODE REVIEW

File: dali/python/nvidia/dali/plugin/paddle.py [hunk 1]
```diff
@@ -72,6 +73,8 @@ def feed_ndarray(dali_tensor, ptr, cuda_stream=None):
                     (if not provided, an internal user stream will be selected)
     """
 
+    if cuda_stream is None:
```
Comment (JanuszL) [inline dali/python/nvidia/dali/plugin/paddle.py:76]: How about moving this to L326 (like in the case of pytorch)?
Please also remove `fluid.core._cuda_synchronize(pd_gpu_place)` which was supposed to be a workaround for the mentioned problem.
Side question - which paddle paddle version is needed to support this maybe there should an if for the version?
And you can update qa/setup_packages.py to make DALI tests use the more recent paddlepaddle-gpu version.

**Comment by JanuszL** [comment]
Hi @sneaxiy,

Thank you for the fix. Would you mind checking my comment?

File: dali/python/nvidia/dali/plugin/paddle.py (hunk 1 above)
Comment (sneaxiy) [reply dali/python/nvidia/dali/plugin/paddle.py:76]: Done. Thanks for your suggestions!

**Comment by JanuszL** [comment]
!build

**Comment by JanuszL** [comment]
Hi @sneaxiy,

Thank you for the update. Let me see if it passes our CI.

**Review by JanuszL — APPROVED** [review]

**Review by jantonguirao — APPROVED** [review]

**Comment by JanuszL** [comment]
@sneaxiy - thank you again for your contribution.

**Pull request merged**